pkg/torcx: introduce an image manifest for propagation phase

lucab opened the pull request:
This introduces a new `.torcx/manifest.json` to specify which assets contained in the image needs to be propagated to the host system. It currently supports propagating binaries (to `/var/run/torcx/bin`) and systemd service units (to runtime `/run/systemd/system`).

Closes: #29

lucab commented:
Updated, @squeed PTAL.

squeed commented on examples/image-manifest.json:
```diff
@@ -0,0 +1,12 @@
+      ],
+      "service": [
+          "/lib/systemd/system/cups.service",
+          "/lib/systemd/system/printer.target.wants/"
```
Does this imply the whole directory is copied?

squeed replied:
Nevermind, I see it!

lucab replied:
Just for reference: yes, the whole directory is copied. If it already exists, its content is merged in the existing one without overwriting existing assets.

squeed commented on pkg/torcx/propagate.go:
```diff
@@ -15,94 +15,197 @@
-				"src": srcPath,
-				"dst": dstPath,
-			}).Debug("installing unit file")
-			return os.Symlink(srcPath, dstPath)
```
Hmm, am I missing something? How do the .wants symlinks get created?

lucab replied:
No, you are right I forgot a case here.

squeed commented on pkg/torcx/propagate.go:
```diff
@@ -15,94 +15,197 @@
+
+// flattenBinAssets propagates a single binary or a directory of binaries,
+// flattening all intermediate directories.
+func flattenBinAsset(applyCfg *ApplyConfig, binDir string, asset string) error {
```
Bikeshedding: The name "flatten" is a bit confusing to me. I associate it with collection code, and don't expect it to have side-effects. Maybe call it `propagateBinAsset`?

lucab replied:
`propagateBins` is already the parent caller of this, while the purpose of this is to flatten any specified assets directory and just symlink every single binary in it. Perhaps `symlinkBinAsset`?

squeed commented:
Seems reasonable. For docker, we'll need the ability to propagate `systemd-networkd` files as well, if you want to think about that.

lucab commented:
Second iteration, PTAL:
 * renamed `flatten*` to `symlink*`
 * fixed symlinked-units propagation

I'm keeping network and udev assets out of this on purpose, I'll attack them in a later after diggin a bit in networkd and udevd setup in CL.

squeed commented:
LGTM, awesome!
